File: skills/ecological-data-foundation/scripts/clean_occurrences.py
#!/usr/bin/env python3
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

"""
clean_occurrences.py
Standard occurrence cleaning pipeline using Python.
Usage: python clean_occurrences.py <input_csv> <output_dir>
Requires: pandas, geopandas, shapely
"""

import logging
import sys
from datetime import datetime
from pathlib import Path

SKILL_NAME = "ecological-data-foundation"
_LOG_DIR   = Path("logs")
_LOG_DIR.mkdir(parents=True, exist_ok=True)
_log_file  = _LOG_DIR / f"skill_{SKILL_NAME}_{datetime.now().strftime('%Y%m%d_%H%M%S')}.log"
logging.basicConfig(
    level=logging.INFO,
    format="[%(asctime)s] [%(levelname)s] [" + SKILL_NAME + "] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
    handlers=[
        logging.StreamHandler(sys.stdout),
        logging.FileHandler(_log_file, encoding="utf-8"),
    ],
)
logger = logging.getLogger(SKILL_NAME)

def log_step(n: int, desc: str) -> None:
    logger.info("-- STEP %d: %s", n, desc)

def log_decision(var: str, val, why: str) -> None:
    logger.info("DECISION | %s = %s | %s", var, val, why)


import os
import hashlib

import pandas as pd
import numpy as np


def load(path: str) -> pd.DataFrame:
    try:
        df = pd.read_csv(path, low_memory=False)
        logger.info("Carregadas %d linhas, %d colunas de %s", len(df), len(df.columns), path)
        return df
    except FileNotFoundError:
        logger.error(
            "Input nao encontrado: %s\n  Causa provavel: arquivo nao gerado pelo passo anterior.\n  Verifique a saida de: ecological-data-foundation (download_from_gbif)\n  Skill anterior: ecological-data-foundation",
            path,
        )
        sys.exit(1)
    except Exception as e:
        logger.error(
            "Falha ao ler CSV de entrada '%s': %s\n  Causa provavel: arquivo corrompido ou formato invalido.\n  Skill anterior: ecological-data-foundation",
            path, e,
        )
        sys.exit(1)


def check_required_cols(df: pd.DataFrame, required: list) -> None:
    missing = [c for c in required if c not in df.columns]
    if missing:
        logger.error(
            "Colunas obrigatorias ausentes: %s\n  Causa provavel: CSV gerado por fonte diferente ou schema alterado.\n  Skill anterior: ecological-data-foundation",
            missing,
        )
        raise ValueError(f"Missing required columns: {missing}")
    logger.info("Todas as colunas obrigatorias presentes: %s", required)


def flag_coordinate_issues(df: pd.DataFrame,
                           lat_col="decimalLatitude",
                           lon_col="decimalLongitude") -> pd.DataFrame:
    df = df.copy()
    df["QA_status"] = "OK"
    lat = pd.to_numeric(df[lat_col], errors="coerce")
    lon = pd.to_numeric(df[lon_col], errors="coerce")

    # Invalid range
    mask_range = (lat.abs() > 90) | (lon.abs() > 180)
    df.loc[mask_range, "QA_status"] = "COORD_OUT_OF_RANGE"

    # Zero coordinates
    mask_zero = (lat == 0) & (lon == 0)
    df.loc[mask_zero, "QA_status"] = "COORD_ZERO"

    # Missing coords
    mask_na = lat.isna() | lon.isna()
    df.loc[mask_na, "QA_status"] = "MISSING_COORDS"

    n_range = int(mask_range.sum())
    n_zero  = int(mask_zero.sum())
    n_na    = int(mask_na.sum())

    logger.info(
        "Problemas de coordenadas — Fora do intervalo: %d | Zero: %d | Ausentes: %d",
        n_range, n_zero, n_na,
    )
    if n_range > 0:
        logger.warning("Registros com coordenadas fora do intervalo valido: %d", n_range)
    if n_zero > 0:
        logger.warning("Registros com coordenadas zero (0,0): %d — possivelmente erros de digitacao", n_zero)
    if n_na > 0:
        logger.warning("Registros sem coordenadas (NA): %d", n_na)
    return df


def remove_exact_duplicates(df: pd.DataFrame,
                             cols=("scientificName","decimalLatitude","decimalLongitude","eventDate")
                             ) -> pd.DataFrame:
    cols_present = [c for c in cols if c in df.columns]
    log_decision(
        "dedup_cols",
        cols_present,
        "combinacao padrao para identificar duplicatas espaciotemporais",
    )
    n_before = len(df)
    df = df.drop_duplicates(subset=cols_present, keep="first")
    n_removed = n_before - len(df)
    if n_removed > 0:
        logger.warning("Duplicatas exatas removidas: %d", n_removed)
    else:
        logger.info("Nenhuma duplicata exata encontrada.")
    return df


def check_temporal(df: pd.DataFrame, date_col="eventDate") -> pd.DataFrame:
    if date_col not in df.columns:
        logger.warning("Coluna '%s' ausente — verificacao temporal pulada.", date_col)
        return df
    dates = pd.to_datetime(df[date_col], errors="coerce")
    future = dates > pd.Timestamp.now()
    n_future = int(future.sum())
    df.loc[future & (df["QA_status"] == "OK"), "QA_status"] = "DATE_FUTURE"
    if n_future > 0:
        logger.warning("Registros com datas futuras sinalizados: %d", n_future)
    else:
        logger.info("Nenhum registro com data futura encontrado.")
    return df


def write_outputs(df: pd.DataFrame, output_dir: Path) -> None:
    output_dir.mkdir(parents=True, exist_ok=True)
    clean   = df[df["QA_status"] == "OK"]
    flagged = df[df["QA_status"] != "OK"]

    try:
        clean.to_csv(output_dir / "data_clean.csv", index=False)
        logger.info("Gravado: %s", output_dir / "data_clean.csv")
        flagged.to_csv(output_dir / "flagged_records.csv", index=False)
        logger.info("Gravado: %s", output_dir / "flagged_records.csv")
    except OSError as e:
        logger.error(
            "Falha ao gravar arquivos de saida em '%s': %s\n  Causa provavel: sem permissao de escrita no diretorio.\n  Skill anterior: ecological-data-foundation",
            output_dir, e,
        )
        raise

    # QA report
    flag_counts = flagged["QA_status"].value_counts().to_dict()
    lines = [
        "# QA Report — Occurrence Cleaning",
        f"- Raw records: {len(df):,}",
        f"- Clean records: {len(clean):,}",
        f"- Flagged records: {len(flagged):,}",
        "",
        "## Flag Counts",
    ] + [f"- `{k}`: {v}" for k, v in flag_counts.items()]

    try:
        (output_dir / "qa_report.md").write_text("\n".join(lines))
        logger.info("Gravado: %s", output_dir / "qa_report.md")
    except OSError as e:
        logger.error(
            "Falha ao gravar relatorio QA: %s\n  Causa provavel: sem permissao de escrita.\n  Skill anterior: ecological-data-foundation",
            e,
        )
        raise

    if len(clean) < 30:
        logger.warning(
            "Apenas %d registros limpos apos todas as filtragens. SDMs requerem >= 30 registros confiaveis.",
            len(clean),
        )

    logger.info("Concluido. Limpos: %d | Sinalizados: %d", len(clean), len(flagged))
    logger.info("Saidas gravadas em: %s", output_dir)


def main():
    logger.info("Script: clean_occurrences.py | Skill: %s", SKILL_NAME)

    input_file = sys.argv[1] if len(sys.argv) > 1 else "data/raw/occurrences.csv"
    output_dir = Path(sys.argv[2]) if len(sys.argv) > 2 else Path("data/processed")

    logger.info("Input file : %s", input_file)
    logger.info("Output dir : %s", output_dir)

    # Input precondition check
    if not Path(input_file).exists():
        logger.error(
            "Input nao encontrado: %s\n  Causa provavel: arquivo nao gerado pelo passo anterior.\n  Verifique a saida de: ecological-data-foundation (download_from_gbif)\n  Skill anterior: ecological-data-foundation",
            input_file,
        )
        sys.exit(1)

    log_decision("input_file", input_file, "caminho passado como argv[1] ou padrao")
    log_decision("output_dir", str(output_dir), "caminho passado como argv[2] ou padrao")

    log_step(1, "Carregar dados brutos de ocorrencias")
    df = load(input_file)

    if len(df) == 0:
        logger.warning("Arquivo de entrada nao contem registros: %s", input_file)

    log_step(2, "Verificar colunas obrigatorias")
    try:
        check_required_cols(df, ["decimalLatitude", "decimalLongitude"])
    except ValueError as e:
        logger.error("Verificacao de colunas falhou: %s", e)
        sys.exit(1)

    log_step(3, "Sinalizar problemas de coordenadas")
    log_decision(
        "cc_tests",
        "out_of_range,zero,missing",
        "verificacoes basicas de qualidade de coordenadas sem dependencias externas",
    )
    try:
        df = flag_coordinate_issues(df)
    except Exception as e:
        logger.error(
            "Falha na sinalizacao de coordenadas: %s\n  Causa provavel: colunas de coordenadas com tipos inesperados.\n  Skill anterior: ecological-data-foundation",
            e,
        )
        raise

    log_step(4, "Remover duplicatas exatas")
    try:
        df = remove_exact_duplicates(df)
    except Exception as e:
        logger.error(
            "Falha ao remover duplicatas: %s\n  Causa provavel: colunas de deduplicacao ausentes ou mal tipadas.\n  Skill anterior: ecological-data-foundation",
            e,
        )
        raise

    log_step(5, "Verificar datas futuras")
    try:
        df = check_temporal(df)
    except Exception as e:
        logger.error(
            "Falha na verificacao temporal: %s\n  Causa provavel: coluna eventDate com formato inesperado.\n  Skill anterior: ecological-data-foundation",
            e,
        )
        raise

    log_step(6, "Escrever arquivos de saida e relatorio QA")
    write_outputs(df, output_dir)


if __name__ == "__main__":
    main()

File: skills/ecological-data-foundation/scripts/clean_occurrences.R
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

# Usage: Rscript clean_occurrences.R <raw_occurrences.csv> <output_dir> [country_code]

# ── Inline logger ─────────────────────────────────────────────────────────────
SKILL_NAME <- "ecological-data-foundation"
.log_ts  <- function() format(Sys.time(), "[%Y-%m-%d %H:%M:%S]")
log_info <- function(...) message(.log_ts(), " [INFO]  ", sprintf(...))
log_warn <- function(...) message(.log_ts(), " [WARN]  ", sprintf(...))
log_error<- function(...) message(.log_ts(), " [ERROR] ", sprintf(...))
log_step <- function(n, d) log_info("-- STEP %d: %s", n, d)
log_decision <- function(v, val, why) log_info("DECISION | %s = %s | %s", v, val, why)
dir.create("logs", recursive=TRUE, showWarnings=FALSE)

# Standard occurrence cleaning pipeline
# Usage: Rscript clean_occurrences.R <input_csv> <output_dir>
# Requires: dplyr, readr, CoordinateCleaner, taxize, janitor

suppressPackageStartupMessages({
  library(dplyr)
  library(readr)
  library(CoordinateCleaner)
  library(janitor)
})

args <- commandArgs(trailingOnly = TRUE)
input_file  <- ifelse(length(args) >= 1, args[1], "data/raw/occurrences.csv")
output_dir  <- ifelse(length(args) >= 2, args[2], "data/processed")

log_info("Script: clean_occurrences.R | Skill: %s", SKILL_NAME)
log_info("Input file : %s", input_file)
log_info("Output dir : %s", output_dir)

# ── Input precondition check ──────────────────────────────────────────────────
if (!file.exists(input_file)) {
  log_error(
    "Input nao encontrado: %s\nCausa provavel: arquivo nao gerado pelo passo anterior.\nVerifique a saida de: ecological-data-foundation (download_from_gbif)\nSkill anterior: ecological-data-foundation",
    input_file
  )
  stop("Missing: ", input_file)
}

dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

log_decision("input_file", input_file, "caminho passado como args[1] ou padrao")
log_decision("output_dir", output_dir, "caminho passado como args[2] ou padrao")

# ── 1. Ingest ──────────────────────────────────────────────────────────────
log_step(1, "Ingerir dados brutos de ocorrencias")
tryCatch({
  raw <- read_csv(input_file, show_col_types = FALSE) |>
    clean_names()
  log_info("Registros brutos lidos: %d | Colunas: %d", nrow(raw), ncol(raw))
}, error = function(e) {
  log_error(
    "Falha ao ler CSV de entrada: %s\nCausa provavel: arquivo corrompido ou nao e CSV valido.\nVerifique: %s\nSkill anterior: ecological-data-foundation",
    conditionMessage(e), input_file
  )
  stop(e)
})

if (nrow(raw) == 0) {
  log_warn("Arquivo de entrada nao contem registros: %s", input_file)
}

# ── 2. Require minimum columns ─────────────────────────────────────────────
log_step(2, "Verificar colunas obrigatorias")
tryCatch({
  required_cols <- c("decimal_latitude", "decimal_longitude", "species")
  missing_req <- setdiff(required_cols, names(raw))
  if (length(missing_req) > 0) {
    log_error(
      "Colunas obrigatorias ausentes: %s\nCausa provavel: CSV gerado por fonte diferente ou com nomes de colunas alterados.\nVerifique o schema do arquivo: %s\nSkill anterior: ecological-data-foundation",
      paste(missing_req, collapse = ", "), input_file
    )
    stop("Missing required columns: ", paste(missing_req, collapse = ", "))
  }
  log_info("Todas as colunas obrigatorias presentes: %s", paste(required_cols, collapse = ", "))
}, error = function(e) {
  log_error("Falha na verificacao de colunas: %s", conditionMessage(e))
  stop(e)
})

# ── 3. Remove records with missing coordinates ─────────────────────────────
log_step(3, "Remover registros sem coordenadas e converter para numerico")
tryCatch({
  n_before_na <- nrow(raw)
  raw <- raw |>
    filter(!is.na(decimal_latitude), !is.na(decimal_longitude)) |>
    mutate(
      decimal_latitude  = as.numeric(decimal_latitude),
      decimal_longitude = as.numeric(decimal_longitude)
    )
  n_removed_na <- n_before_na - nrow(raw)
  if (n_removed_na > 0) {
    log_warn("Registros removidos por coordenadas ausentes: %d", n_removed_na)
  } else {
    log_info("Nenhum registro removido por coordenadas ausentes.")
  }
}, error = function(e) {
  log_error(
    "Falha ao filtrar coordenadas ausentes: %s\nCausa provavel: tipos de coluna inesperados no CSV.\nVerifique: %s\nSkill anterior: ecological-data-foundation",
    conditionMessage(e), input_file
  )
  stop(e)
})

if (nrow(raw) < 30) {
  log_warn("Poucos registros apos remocao de NAs (%d). Minimo recomendado para SDM: 30.", nrow(raw))
}

# ── 4. Coordinate cleaning ─────────────────────────────────────────────────
log_step(4, "Limpeza de coordenadas com CoordinateCleaner")
log_decision(
  "cc_tests",
  "capitals,centroids,equal,gbif,institutions,validity,zeros",
  "conjunto padrao de testes para detectar registros suspeitos"
)
tryCatch({
  flags <- clean_coordinates(
    x       = raw,
    lon     = "decimal_longitude",
    lat     = "decimal_latitude",
    species = "species",
    tests   = c("capitals", "centroids", "equal", "gbif",
                "institutions", "validity", "zeros"),
    verbose = FALSE
  )

  clean    <- raw[flags$.summary, ] |> mutate(QA_status = "OK")
  flagged  <- raw[!flags$.summary, ] |>
    mutate(QA_status = apply(
      flags[!flags$.summary, grep("^\\.", names(flags))], 1,
      function(r) paste(names(r)[!r], collapse = "|")
    ))

  log_info("Registros limpos: %d | Registros sinalizados: %d", nrow(clean), nrow(flagged))
  if (nrow(flagged) > 0) {
    log_warn("%d registros sinalizados pelo CoordinateCleaner (%.1f%% do total).",
             nrow(flagged), 100 * nrow(flagged) / nrow(raw))
  }
}, error = function(e) {
  log_error(
    "Falha na limpeza de coordenadas (CoordinateCleaner): %s\nCausa provavel: dados malformados ou pacote CoordinateCleaner nao instalado.\nVerifique: install.packages('CoordinateCleaner')\nSkill anterior: ecological-data-foundation",
    conditionMessage(e)
  )
  stop(e)
})

# ── 5. Remove exact duplicates ─────────────────────────────────────────────
log_step(5, "Remover duplicatas exatas")
tryCatch({
  n_before <- nrow(clean)
  clean <- clean |> distinct(species, decimal_latitude, decimal_longitude,
                              event_date, .keep_all = TRUE)
  n_dup <- n_before - nrow(clean)
  log_decision(
    "dedup_cols",
    "species,decimal_latitude,decimal_longitude,event_date",
    "combinacao padrao para identificar duplicatas espaciotemporais"
  )
  if (n_dup > 0) {
    log_warn("Duplicatas exatas removidas: %d", n_dup)
  } else {
    log_info("Nenhuma duplicata exata encontrada.")
  }
}, error = function(e) {
  log_error(
    "Falha ao remover duplicatas: %s\nCausa provavel: coluna event_date ausente ou mal formatada.\nVerifique o schema do CSV.\nSkill anterior: ecological-data-foundation",
    conditionMessage(e)
  )
  stop(e)
})

if (nrow(clean) < 30) {
  log_warn(
    "Apenas %d registros limpos apos todas as filtragens. SDMs requerem >= 30 registros confiaveis.",
    nrow(clean)
  )
}

# ── 6. Write outputs ───────────────────────────────────────────────────────
log_step(6, "Escrever arquivos de saida")
tryCatch({
  write_csv(clean,   file.path(output_dir, "data_clean.csv"))
  write_csv(flagged, file.path(output_dir, "flagged_records.csv"))
  log_info("Gravado: %s", file.path(output_dir, "data_clean.csv"))
  log_info("Gravado: %s", file.path(output_dir, "flagged_records.csv"))
}, error = function(e) {
  log_error(
    "Falha ao gravar arquivos de saida: %s\nCausa provavel: sem permissao de escrita em '%s'.\nVerifique permissoes do diretorio.\nSkill anterior: ecological-data-foundation",
    conditionMessage(e), output_dir
  )
  stop(e)
})

# ── 7. QA report ──────────────────────────────────────────────────────────
log_step(7, "Gerar relatorio de QA")
tryCatch({
  report <- c(
    "# QA Report — Occurrence Cleaning",
    "",
    paste("- Input file:", input_file),
    paste("- Raw records:", nrow(raw) + nrow(flagged)),
    paste("- Exact duplicates removed:", n_dup),
    paste("- Records flagged by CoordinateCleaner:", nrow(flagged)),
    paste("- Clean records written:", nrow(clean)),
    "",
    "## Flag Counts",
    ""
  )

  flag_cols <- grep("^\\.", names(flags), value = TRUE)
  for (fc in flag_cols) {
    n_fail <- sum(!flags[[fc]], na.rm = TRUE)
    if (n_fail > 0) report <- c(report, paste0("- `", fc, "`: ", n_fail))
  }

  writeLines(report, file.path(output_dir, "qa_report.md"))
  log_info("Gravado: %s", file.path(output_dir, "qa_report.md"))
}, error = function(e) {
  log_error(
    "Falha ao gerar relatorio QA: %s\nCausa provavel: problema ao escrever no diretorio '%s'.\nSkill anterior: ecological-data-foundation",
    conditionMessage(e), output_dir
  )
  stop(e)
})

log_info("Concluido. Registros limpos: %d | Sinalizados: %d", nrow(clean), nrow(flagged))
